### DISCUSSION TITLE
How to handle "variants shared/static not matching the actual installation"?

### DISCUSSION THREAD

**Discussion by climbfuji** [spack/spack#29349] [Q&A]
This is a problem that has caused headache for some time.

**Example:**
- spack stack builds fine on a particular machine
- adding parallel-netcdf and enabling it in netcdf-c, netcdf-fortran, parallelio, ...
- all of a sudden netcdf-fortran and nccmp don't build anymore, completely obscure error message that the netcdf-c library can't be found

**Reason:**
- parallel-netcdf for some reason on this particular machine builds static libraries only
    - the package still builds although the default is to request shared and static libraries
- netcdf-c figures this out and builds static libraries only
    - but spack thinks it build shared libraries because +shared is in the specs
    - therefore calls to get the netcdf-c library linker flags are wrong
- netcdf-fortran doesn't build because it can't find the netcdf-c dependencies (and similar for nccmp etc)

**Solution?**

To me, the correct, best, but unrealistic solution iss that every package
- is configured to understand the static and shared variants and "reply" to them
    - if a user requests a build type that this package doesn't do, fail
    - if a user requests both static and shared and this package can't do both, fail
    - if a user requests neiter static nor shared, fail
    - if a requested build type can't be used on a particular machine, fail
- ideally can be queried by spack post-install so that spack knows which versions are actually available

Since this is very much impossible to achive, is there a way to inspect the libraries post install in spack and raise an error if they don't match the build specs?

**Comment by scheibelp** [comment]
You can add a per-package post-install function that would fail the install if expected libraries are not present. In general, we could update the Spack core to apply this check for all packages which include a `+shared`/`+static` variant: I think a general solution (e.g. make sure there is a `.a` for `+static` or `~shared`) would work and in that case users wouldn't have to implement essentially the same check in each package where they care about this.

Such a check would help people detect cases where the actual installation was out of sync with the spec's recorded variants. The response would likely be package specific (e.g. in the case of `parallel-netcdf`, add a `conflicts(+shared, when=...)`.

**Reply by climbfuji** [reply]
Thanks, yes, this sounds like what I was looking for!
